lib/reversetunnel/seek/lease.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package seek
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+// Lease is a per-agent lease; used to manage the number of simultaneously
+// connecting agents.  The SeekPool grants leases when new agents need to
+// be spawned.  Agents release their leases when they halt.
+type Lease struct {
+	GroupHandle
+	once    sync.Once
+	release chan<- struct{}
+	id      uint64
+}
+
+// WithProxy is used to wrap the connection-handling logic of an agent,
+// ensuring that it is run if and only if no other agent is already
+// handling this proxy.
+func (l *Lease) WithProxy(do func(), principals ...string) (did bool) {
+	return l.GroupHandle.WithProxy(do, l.ID(), principals...)
+}
+
+// ID returns the unique ID of this lease.
+func (l *Lease) ID() uint64 {
+	return l.id
+}
+
+// Release releases the lease.
+func (l *Lease) Release() {
+	l.once.Do(func() {
+		select {
+		case l.release <- struct{}{}:
+		case <-l.Done():
+		}
+	})
+}
+
+func (l *Lease) String() string {
+	return fmt.Sprintf("Lease(key=%v,id=%v)", l.Key(), l.id)
+}
+
+type counter struct {
```

Comment:
This can be simplified:
```
type counter uint64

func (c *counter) next() uint64 {
	return atomic.AddUint64((*uint64)(c), 1)
}

...
c := new(counter)
id := c.next()
```

Reply:
Extra API complexity was deliberate.  Bare atomics are too easy to misuse IMO.  Wrapping the pointer allows `counter` to panic on use of zero value, and guarantees correct memory alignment on all platforms (not that we currently support any of the platforms with serious [alignment gotchas](https://golang.org/pkg/sync/atomic/#pkg-note-BUG), but hey).

Reply:
Looks really similar to https://github.com/uber-go/atomic may be adopt it here and for other similar cases? the use case seems to be generic enough

Reply:
Zero value could be made valid and safe for use.
Didn't know about the MIPS alignment problems, that's surprising and totally justifies wrapping with a struct.

How about deleting the `Load` method though? It seems unused right now.

Reply:
> Looks really similar to https://github.com/uber-go/atomic may be adopt it here and for other similar cases?

I like this idea.  We already track `uber-go/atomic` as an indirect dependency anyhow.

> Didn't know about the MIPS alignment problems, that's surprising and totally justifies wrapping with a struct.

Yeah, its the kind of thing that should really be declared at head of the `sync/atomic` docs.